File: cw_hrms/cw_hrms/page/custom_user_dashboar/custom_user_dashboar.py
import frappe
from frappe.utils import nowdate, get_first_day, get_last_day, getdate, add_days, flt

@frappe.whitelist()
def get_user_stats():
    user = frappe.session.user
    employee_doc = frappe.db.get_value("Employee", {"user_id": user}, ["name", "company", "holiday_list"], as_dict=True)
    
    if not employee_doc:
        return {"stats": {}, "attendance_details": [], "leave_allocation": [], "leave_history": []}

    employee = employee_doc.name
    today = getdate(nowdate())
    month_start = getdate(get_first_day(today))
    
    # year_start এবং year_end তৈরি করা হলো (এরর ফিক্স)
    year_start = f"{today.year}-01-01"
    year_end = f"{today.year}-12-31"

    data = {
        "stats": {"present": 0, "absent": 0, "late": 0, "leave": 0, "holiday": 0, "weekend": 0, "home_office": 0},
        "leave_allocation": [],
        "attendance_details": [],
        "leave_history": []
    }

    # ১. এটেনডেন্স ডাটা সংগ্রহ
    attendance_records = frappe.get_all("Attendance", 
        filters={"employee": employee, "attendance_date": ["between", [month_start, today]], "docstatus": 1},
        fields=["attendance_date", "status", "late_entry", "in_time", "out_time", "working_hours"]
    )
    att_dict = {getdate(d.attendance_date): d for d in attendance_records}

    # ২. হলিডে লিস্ট ডাটা সংগ্রহ
    h_list = employee_doc.holiday_list or frappe.db.get_value("Company", employee_doc.company, "default_holiday_list")
    holiday_dict = {}
    if h_list:
        holidays = frappe.db.sql("""SELECT holiday_date, description, weekly_off FROM `tabHoliday` 
                                    WHERE parent = %s AND holiday_date BETWEEN %s AND %s""", 
                                 (h_list, month_start, today), as_dict=True)
        holiday_dict = {getdate(h.holiday_date): h for h in holidays}

    # ৩. লুপ প্রসেসিং (১ তারিখ থেকে আজ পর্যন্ত)
    curr_date = month_start
    temp_details = []

    while curr_date <= today:
        status_to_show = ""
        in_t, out_t, work_h = None, None, 0
        
        # Priority 1: Holiday List
        if curr_date in holiday_dict:
            h_info = holiday_dict[curr_date]
            if curr_date in att_dict and att_dict[curr_date].status == "Present":
                att = att_dict[curr_date]
                status_to_show = "Present"
                in_t, out_t, work_h = att.in_time, att.out_time, att.working_hours
                data["stats"]["present"] += 1
                if att.late_entry: data["stats"]["late"] += 1
            else:
                if h_info.weekly_off:
                    status_to_show = "Weekend"
                    data["stats"]["weekend"] += 1
                else:
                    status_to_show = h_info.description or "Holiday"
                    data["stats"]["holiday"] += 1 # এখানে ফিক্স করা হয়েছে

        # Priority 2: Attendance Database
        elif curr_date in att_dict:
            att = att_dict[curr_date]
            status_to_show = att.status
            in_t, out_t, work_h = att.in_time, att.out_time, att.working_hours
            
            if status_to_show == "Present": data["stats"]["present"] += 1
            elif status_to_show == "On Leave": data["stats"]["leave"] += 1
            elif status_to_show == "Work From Home": data["stats"]["home_office"] += 1
            elif status_to_show == "Absent": data["stats"]["absent"] += 1
            if att.late_entry: data["stats"]["late"] += 1

        # Priority 3: Absent
        else:
            status_to_show = "Absent"
            data["stats"]["absent"] += 1

        temp_details.append({
            "attendance_date": curr_date,
            "status": status_to_show,
            "in_time": in_t,
            "out_time": out_t,
            "working_hours": work_h
        })
        curr_date = add_days(curr_date, 1)

    data["attendance_details"] = sorted(temp_details, key=lambda x: x['attendance_date'], reverse=True)
    
    # ৪. লিভ ডাটা (Allocation & History)
    # Allocation-এ 'unused_leaves' ও আনা হয়েছে যাতে অবশিষ্ট ছুটি দেখা যায়
    data["leave_allocation"] = frappe.get_all("Leave Allocation",
        filters={"employee": employee, "docstatus": 1, "from_date": ["<=", today], "to_date": [">=", today]},
        fields=["leave_type", "total_leaves_allocated as total", "unused_leaves"])
    
    data["leave_history"] = frappe.get_all("Leave Application",
        filters={"employee": employee, "status": "Approved", "from_date": ["between", [year_start, year_end]]},
        fields=["leave_type", "from_date", "to_date", "total_leave_days"],
        order_by="from_date desc")

    return data

File: cw_hrms/cw_hrms/page/custom_user_dashboar/custom_user_dashboar.js
frappe.pages['custom-user-dashboar'].on_page_load = function(wrapper) {
    let page = frappe.ui.make_app_page({
        parent: wrapper,
        title: __('User Dashboard'),
        single_column: true
    });

    frappe.call({
        method: 'cw_hrms.cw_hrms.page.custom_user_dashboar.custom_user_dashboar.get_user_stats',
        callback: function(r) {
            if (r.message) {
                render_dashboard_html(page, r.message);
            }
        }
    });
};

function render_dashboard_html(page, data) {
    // stats অবজেক্ট বা ডিফল্ট খালি অবজেক্ট
    let stats = data.stats || {};
    
    let html = `
    <div class="container-fluid p-4" style="background-color: #f8f9fa;">
        <div class="row">
            <div class="col-md-8">
                <div class="row">
                    ${createCard("Present", stats.present, "#28a745")}
                    ${createCard("Absent", stats.absent, "#dc3545")}
                    ${createCard("Late", stats.late, "#ffc107")}
                </div>
                <div class="row mt-3">
                    ${createCard("Leave", stats.leave, "#6f42c1")}
                    ${createCard("Weekend/Holiday", stats.weekend || stats.holiday, "#17a2b8")}
                    ${createCard("Home Office", stats.home_office, "#fd7e14")}
                </div>
            </div>

            <div class="col-md-4">
                <div class="card shadow-sm mb-3 border-0">
                    <div class="card-header bg-white font-weight-bold">Leave Allocation</div>
                    <div class="p-2">
                        ${(data.leave_allocation && data.leave_allocation.length) ? data.leave_allocation.map(l => `
                            <div class="d-flex justify-content-between p-2 border-bottom">
                                <span class="text-muted">${l.leave_type}</span>
                                <span class="badge badge-primary badge-pill">${l.total}</span>
                            </div>
                        `).join('') : '<p class="text-center p-2 text-muted">No Allocation</p>'}
                    </div>
                </div>

                <div class="card shadow-sm border-0">
                    <div class="card-header bg-white font-weight-bold">Quick Links</div>
                    <div class="card-body p-2">
                        <button class="btn btn-outline-primary btn-sm btn-block mb-2" onclick="frappe.set_route('List', 'Leave Application', 'List')">Leave Application</button>
                        <button class="btn btn-outline-secondary btn-sm btn-block" onclick="frappe.set_route('List', 'Employee Checkin', 'List')">Employee Checkin</button>
                    </div>
                </div>
            </div>
        </div>

        <div class="row mt-5">
            <div class="col-md-6">
                <div class="card shadow-sm border-0">
                    <div class="card-header font-weight-bold bg-white border-0">Attendance Summary (Monthly)</div>
                    <div class="table-responsive" style="max-height: 400px;">
                        <table class="table table-sm table-hover text-center mb-0">
                            <thead class="bg-light text-muted">
                                <tr><th>SL</th><th>Date</th><th>In</th><th>Out</th><th>Stay</th><th>Status</th></tr>
                            </thead>
                            <tbody>
                                ${data.attendance_details.map((a, i) => {
                                    // স্ট্যাটাস অনুযায়ী ব্যাজ কালার নির্ধারণ
                                    let badge_class = "secondary";
                                    if(a.status === "Present") badge_class = "success";
                                    else if(a.status === "Absent") badge_class = "danger";
                                    else if(a.status === "Weekend" || a.status === "Holiday") badge_class = "info";
                                    else if(a.status === "Work From Home") badge_class = "warning";
                                    else if(a.status === "On Leave") badge_class = "primary";

                                    return `
                                    <tr>
                                        <td>${i+1}</td>
                                        <td>${frappe.datetime.str_to_user(a.attendance_date)}</td>
                                        <td>${a.in_time ? frappe.datetime.get_time(a.in_time) : '--'}</td>
                                        <td>${a.out_time ? frappe.datetime.get_time(a.out_time) : '--'}</td>
                                        <td>${a.working_hours ? flt(a.working_hours).toFixed(2) : '0'}h</td>
                                        <td><span class="badge badge-${badge_class}">${a.status}</span></td>
                                    </tr>`;
                                }).join('')}
                            </tbody>
                        </table>
                    </div>
                </div>
            </div>

            <div class="col-md-6">
                <div class="card shadow-sm border-0">
                    <div class="card-header font-weight-bold bg-white border-0">Leave Summary (Yearly)</div>
                    <div class="table-responsive" style="max-height: 400px;">
                        <table class="table table-sm table-hover text-center mb-0">
                            <thead class="bg-light text-muted">
                                <tr><th>SL</th><th>From</th><th>To</th><th>Days</th><th>Type</th></tr>
                            </thead>
                            <tbody>
                                ${(data.leave_history && data.leave_history.length) ? data.leave_history.map((l, i) => `
                                    <tr>
                                        <td>${i+1}</td>
                                        <td>${frappe.datetime.str_to_user(l.from_date)}</td>
                                        <td>${frappe.datetime.str_to_user(l.to_date)}</td>
                                        <td>${l.total_leave_days}</td>
                                        <td><span class="badge badge-info">${l.leave_type}</span></td>
                                    </tr>
                                `).join('') : '<tr><td colspan="5" class="p-3 text-muted">No records found</td></tr>'}
                            </tbody>
                        </table>
                    </div>
                </div>
            </div>
        </div>
    </div>`;

    page.main.html(html);
}

// কার্ড তৈরির ফাংশন (N/A হ্যান্ডেল করা হয়েছে)
function createCard(label, val, color) {
    let display_val = (val === undefined || val === null || val === 0) ? "N/A" : val;
    
    return `
    <div class="col-md-4">
        <div class="card p-3 text-center shadow-sm border-0" style="border-top: 4px solid ${color}; border-radius: 8px;">
            <div class="text-muted small font-weight-bold uppercase" style="letter-spacing: 0.5px;">${label}</div>
            <div class="h3 font-weight-bold mt-2" style="color: ${color}">${display_val}</div>
        </div>
    </div>`;
}
